realm-jni/build.gradle

```diff
@@ -89,18 +100,91 @@ task checkProperties(group: 'check', description: 'Check the user provided gradl
     }
 }
 
-// TODO: Split this task into smaller subtasks
+ext.coreArchiveFile = project.file("../core-android-${project.coreVersion}.tar.gz")
+ext.coreDir = new File(buildDir, "core-${project.coreVersion}")
+
+def coreDownloaded = false
+
 task downloadCore(group: 'build setup', description: 'Download the latest version of realm core', dependsOn: checkProperties) {
-    outputs.dir "../core-${project.coreVersion}"
+    def isHashCheckingEnabled = {
+        return project.hasProperty('coreSha256Hash') && !project.coreSha256Hash.empty
+    }
+
+    def calcSha256Hash = {File targetFile ->
+        MessageDigest sha = MessageDigest.getInstance("SHA-256");
+        Formatter hexHash = new Formatter()
+        sha.digest(targetFile.bytes).each { b -> hexHash.format('%02x', b) }
+        return hexHash.toString()
+    }
+
+    def shouldDownloadCore = {
+        if (!project.coreArchiveFile.exists()) {
+            return true
+        }
+        if (project.forceDownloadCore) {
+            return true;
+        }
+        if (!isHashCheckingEnabled()) {
+            println "Skipping hash checking(empty \'coreSha256Hash\')."
+            return false
+        }
+
+        def calculatedHash = calcSha256Hash(project.coreArchiveFile)
+        if (project.coreSha256Hash.equalsIgnoreCase(calculatedHash)) {
+            return false
+        }
+
+        println "Existing archive hash mismatch(Expected: ${project.coreSha256Hash.toLowerCase()}" +
```

Comment:
This sounds a bit "dangerous" also "force downloading" isn't really true, as it will also trigger if a new version is available. Rewriting the last sentence to something like "Download new version" would sound better imo